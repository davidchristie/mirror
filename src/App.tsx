import { useEffect, useRef } from "react";

export function App() {
  const videoRef = useRef<HTMLVideoElement>(null);
  useEffect(() => {
    (async () => {
      if (videoRef.current) {
        try {
          const stream = await navigator.mediaDevices.getUserMedia({
            video: true,
          });
          videoRef.current.srcObject = stream;
        } catch (error) {
          console.error(error);
        }
      }
    })();
  }, []);
  return (
    <div className="App">
      <video autoPlay className="App__video" ref={videoRef} />
    </div>
  );
}
