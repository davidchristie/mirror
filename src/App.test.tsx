import { render } from "@testing-library/react";
import { PassThrough } from "stream";
import { App } from "./App";

describe("<App />", () => {
  const originalNavigator = window.navigator;

  let mockGetUserMedia: jest.Mock;

  beforeAll(() => {
    mockGetUserMedia = jest.fn();
    Object.defineProperty(window.navigator, "mediaDevices", {
      value: {
        getUserMedia: mockGetUserMedia,
      },
    });
  });

  afterAll(() => {
    Object.defineProperty(window, "navigator", {
      value: originalNavigator,
    });
  });

  describe("with camera", () => {
    beforeEach(() => {
      mockGetUserMedia.mockResolvedValue(new PassThrough());
    });

    it("renders correctly", () => {
      const result = render(<App />);
      const app = result.container.firstChild;
      expect(app).toHaveClass("App");
      expect(app).toMatchSnapshot();
    });
  });

  describe("without camera", () => {
    beforeEach(() => {
      mockGetUserMedia.mockRejectedValue(new DOMException("Permission denied"));
    });

    it("renders correctly", () => {
      const result = render(<App />);
      const app = result.container.firstChild;
      expect(app).toHaveClass("App");
      expect(app).toMatchSnapshot();
    });
  });
});
