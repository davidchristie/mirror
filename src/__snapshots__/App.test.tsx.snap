// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> with camera renders correctly 1`] = `
<div
  class="App"
>
  <video
    autoplay=""
    class="App__video"
  />
</div>
`;

exports[`<App /> without camera renders correctly 1`] = `
<div
  class="App"
>
  <video
    autoplay=""
    class="App__video"
  />
</div>
`;
